code/game/objects/structures/training_machine.dm

```diff
@@ -0,0 +1,264 @@
+	var/turf/current_turf = get_turf(src)
+	if (!moving || !starting_turf || isspaceturf(current_turf))
+		stop_moving()
+		return
+	if (current_turf == target_position) //We've reached our target turf, now find a new one
+		target_position = null
+	if (!target_position)
+		target_position = find_target_position()
+		if (!target_position)
+			stop_moving("ERROR! Cannot calculate suitable movement path.")
+	var/turf/nextStep = get_step_towards(src, target_position)
+	if (!Move(nextStep))
+		target_position = null //We couldn't move towards the target turf, so find a new target turf
+	if (obj_flags & EMAGGED && world.time > last_attack_time + attack_cooldown)
+		var/mob/living/carbon/target = locate() in view(1, src)
+		if (target?.stat == CONSCIOUS)
+			do_attack_animation(target, null, secret_evil_toolbox)
+			target.apply_damage(secret_evil_toolbox.force, BRUTE, BODY_ZONE_HEAD)
+			playsound(src,'sound/weapons/smash.ogg',50,FALSE)
+			last_attack_time = world.time
+			attack_cooldown = rand(MIN_ATTACK_DELAY,MAX_ATTACK_DELAY)
+
+	addtimer(CALLBACK(src, .proc/do_movement), max(MAX_SPEED - move_speed, 1) // We want to ensure this is never <=0
+
+/obj/structure/training_machine/proc/find_target_position()
+	var/list/turfs = list()
+	for(var/turf/potential_turf in view(range, starting_turf))
+		if (potential_turf.is_blocked_turf() || potential_turf == target_position)
+			continue
+		turfs += potential_turf
+	if (!length(turfs))
+		return
+	return pick(turfs)
+
+/obj/structure/training_machine/proc/handle_density()
+	if(length(buckled_mobs) || pinned_target || (obj_flags & EMAGGED))
+		density = TRUE
+	else
+		density = FALSE
+
+/obj/structure/training_machine/buckle_mob(mob/living/M, force = FALSE, check_loc = TRUE)
+	. = ..()
+	if (pinned_target)
+		return FALSE
+
+/obj/structure/training_machine/post_buckle_mob()
+	handle_density()
+	return ..()
+
+/obj/structure/training_machine/post_unbuckle_mob()
+	handle_density()
+	return ..()
+
+/obj/structure/training_machine/emag_act(mob/user)
+	. = ..()
+	if (obj_flags & EMAGGED)
+		return
+	obj_flags |= EMAGGED
+	secret_evil_toolbox = new(src)
+	to_chat(user, "<span class='warning'>You override the training machine's safety protocols, and activate its realistic combat feature. A toolbox pops out of a slot on the top.</span>")
+	playsound(src, 'sound/machines/click.ogg', 50, TRUE)
+	add_overlay("tm_toolbox")
+	handle_density()
+
+/obj/structure/training_machine/examine(mob/user)
+	. = ..()
+	. += "You notice it has a set of <span class='notice'><b>buckling straps</b></span>"
+	if (obj_flags & EMAGGED)
+		. += "<span class='warning'>It has a dangerous-looking toolbox attached to it...</span>"
+	if (pinned_target)
+		. += "<span class='notice'><b>Alt-Click to remove pinned target</b></span>"
+	. += "<span class='notice'><b>Click to open control interface.</b></span>"
+
+/obj/structure/training_machine/Destroy()
+	. = ..()
+	if (secret_evil_toolbox)
+		QDEL_NULL(secret_evil_toolbox)
```

Comment:
Please put the `Initialize()` and `Destroy()` procs right after the object definition, for better organization.